### PR TITLE
Add oauth1 support

### PR DESCRIPTION
**Pull request by jsantos17**
This adds support for OAuth1. Has only been tested with Twitter's API. I can add tests if there's interest in getting this merged.

### CODE REVIEW

**Comment by jsantos17** [comment]
@bos Hey, sorry to bother you, but can you take a look at this to see if it's mergeable?

**Comment by juanpaucar** [comment]
@bos Hi, i would find these quite useful for  project i'm working on right now, would you mind take a look please?
I think is useful because it makes a lot of sense to have OAuth 1.0 since it's quite popular among many services

**Comment by ExternalReality** [comment]
Hmm, I could actually make use of this... could someone make a fork with this...

**Comment by igraves** [comment]
Bumping this.  Was looking to add in OAuth1 support myself, but it looks like @jsantos17 has done it.  Does any additional work need to be done on integration or is this pretty much ready to roll?  Would appreciate a look if you get a chance, @bos.  Thanks!

File: Network/Wreq/Internal/Types.hs [hunk 1]
```diff
@@ -177,6 +177,7 @@ data Auth = BasicAuth S.ByteString S.ByteString
           | AWSAuth AWSAuthVersion S.ByteString S.ByteString
             -- ^ Amazon Web Services request signing
             -- AWSAuthVersion key secret
+          | OAuth1 S.ByteString S.ByteString S.ByteString S.ByteString
```
Comment (bos) [inline Network/Wreq/Internal/Types.hs]: Please document this field.

File: wreq.cabal [hunk 2]
```diff
@@ -80,6 +80,7 @@ library
   other-modules:
     Network.Wreq.Internal
     Network.Wreq.Internal.AWS
+    Network.Wreq.Internal.OAuth1
```
Comment (bos) [inline wreq.cabal]: The module names in this list are sorted. Please add this in the correct lexicographical position.

File: Network/Wreq/Internal/OAuth1.hs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import Web.Authenticate.OAuth ( signOAuth, newOAuth, oauthConsumerKey
+                              , oauthConsumerSecret, newCredential)
+
+signRequest :: Auth -> Request -> IO Request
+signRequest (OAuth1 consumerToken consumerSecret token tokenSecret) requestToSign = signOAuth app creds requestToSign
+  where
+    app = newOAuth { oauthConsumerKey = consumerToken, oauthConsumerSecret = consumerSecret }
+    creds = newCredential token tokenSecret
+
+signRequest _ requestToSign = return requestToSign
```
Comment (bos) [inline Network/Wreq/Internal/OAuth1.hs]: I don't understand the use of a wildcard pattern here. This suggests it needs a comment.

**Comment by bos** [comment]
Sorry for the slow response; I was on vacation. You'll see that I have a few comments, but otherwise this looks good.

File: Network/Wreq/Internal.hs [hunk 4]
```diff
@@ -117,8 +118,10 @@ prepare modify opts url = do
     signRequest = maybe return f $ auth opts
       where
         f (AWSAuth versn key secret) = AWS.signRequest versn key secret
```
Comment (ret) [inline Network/Wreq/Internal.hs]: [bos wrote] While you're at it, please also fix that API mistake for the AWS code in a separate commit

We're not passing the auth record to AWS.signRequest, but only the constituent parts. What could I do better here?

**Comment by jsantos17** [comment]
@bos thanks for the review. I'll get the changes done ASAP :)

**Comment by jsantos17** [comment]
@bos I have made the requested changes. However, I'm not sure what you mean about the AWS API error, maybe @ret knows more about that?

**Comment by jsantos17** [comment]
@bos any chance of getting this merged?

**Comment by jpvillaisaza** [comment]
:+1:

**Comment by CristhianMotoche** [comment]
:+1:

**Comment by bos** [comment]
Applied, thanks!

**Pull request merged**